Source/Particles/WarpXParticleContainer.cpp

```diff
@@ -354,12 +354,33 @@ WarpXParticleContainer::DepositCurrent (WarpXParIter& pti,
     // Jx, Jy and Jz have the same number of guard cells, hence it is sufficient to check for Jx
     const amrex::IntVect range = jx->nGrowVect() - shape_extent;
 #endif
+
+    const std::array<Real,3>& dx = WarpX::CellSize(std::max(depos_lev,0));
+
+    // what is the length of the smallest cell edge?
+#if defined(WARPX_DIM_3D)
+    const Real min_dx = std::min({dx[0],dx[1],dx[2]});
+#elif defined(WARPX_DIM_XZ) || defined(WARPX_DIM_RZ)
+    const Real min_dx = std::min(dx[0],dx[2]);
+#else
+    const Real min_dx = dx[2];
+#endif
+
+    // dt shift would be out-of-range?
+    // We also deposit the current for diagnostics reasons, e.g., in ES simulations that have no CFL
+    // Esirkepov's current deposition algorithm assumes that dt*c does not cross a cell boundary,
+    // let's ensure that.
+    if (WarpX::current_deposition_algo == CurrentDepositionAlgo::Esirkepov) {
+        WARPX_ALWAYS_ASSERT_WITH_MESSAGE(dt * PhysConst::c < min_dx,
```

Comment:
Can this be an `AMREX_ASSERT_WITH_MESSAGE` instead? The actual `dt*PhysConst::c < min_dx` is a stronger constraint than really necessary - would only be needed for particles traveling at the speed of light. I frequently use Esirkepov deposition with the hybrid solver (since it is faster than direct current deposition - not sure why - and less noisy).
The message might also be less rigid - something like "Time step could be too large for Esirkepov..."

Reply:
> [Esirkepov] is faster than direct current deposition

That is unusual - on which platform, with/without sorting, with/without shared memory? :-o
Let's discuss this also with @atmyers in our next dev meeting.

I think that is the primary thing we need to solve.

> would only be needed for particles traveling at the speed of light

That is true, but what guarantee do we have if a particle just reached the velocity that moves more than one cell - and then starts to create memory violations in a dynamic scenario?

I would add a runtime option to skip the test if this helps you, but it seems dangerous to me.

Reply:
Moved to #4364

Reply:
Thanks for the reply. A runtime option to skip the assert would be fine. There is also currently a warning raised if a user chooses to use Esirkepov with the hybrid solver saying that if a particle gains too much speed a segfault will occur in the current deposition. It is not great for ensuring smooth code execution but it at least gives the user a starting place to look if they find that their simulation crashes.

An alternative option I have seen used is to simply limit particle speeds - using the same expression you have to calculate the maximum allowable speed and if a particle is pushed so that it would exceed that speed its speed is set equal to the maximal allowed speed. This obviously violates energy conservation but it prevents the code from crashing. I guess a different strategy could be to set the speed of light to the maximum supported speed in which case the speed limit will be naturally applied through the relativistic corrections... All of that is just for interest sake, I'm not suggesting we do that.

Regarding Esirkepov being faster than direct deposition. I see that with few GPU runs but with many particles, using the hybrid solver. For example, a simulation using 2 Perlmutter GPUs with 100 x 100 x 184 grid size and 400M particles achieves ~1.4 steps per second with Esirkepov but only ~1.15 steps per second with direct deposition. I'll add this to the dev meeting agenda for next week.